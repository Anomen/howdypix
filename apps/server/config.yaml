emailSender:
  email: sender@vickev.com
  name: Howdypix
users:
  - email: dev@vickev.com
    name: Developer Foobar
  - email: albert@vickev.com
    name: Albert Foobar
thumbnailsDir: ./photos
photoDirs:
  main: ./photos/main
  second: ./photos/landscapes
