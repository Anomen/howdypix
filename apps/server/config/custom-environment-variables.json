{
  "app": {
    "rabbitMQ": {
      "url": "RABBITMQ_URL"
    },
    "api": {
      "baseUrl": "API_BASE_URL",
      "port": "API_PORT"
    },
    "webapp": {
      "baseUrl": "WEBAPP_BASE_URL"
    },
    "smtp": {
      "host": "SMTP_HOST",
      "port": "SMTP_PORT",
      "user": "SMTP_USER",
      "password": "SMTP_PASSWORD",
      "tls": "SMTP_TLS"
    }
  }
}
