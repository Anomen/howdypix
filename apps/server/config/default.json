{
  "serverWs": {
    "port": 8998,
    "allowOrigins": ["localhost"]
  },
  "serverApollo": {
    "port": 4005
  }
}
