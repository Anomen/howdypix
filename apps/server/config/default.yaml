#####################################################
# User settings
#####################################################
user:
  emailSender:
    email: sender@vickev.com
    name: Howdypix
  users:
    - email: dev@vickev.com
      name: Developer Foobar
    - email: albert@vickev.com
      name: Albert Foobar
  thumbnailsDir: ./photos
  photoDirs:
    main: ./photos/main
    second: ./photos/landscapes

#####################################################
# Application settings
#####################################################
app:
  database:
    path: database.sqlite

  rabbitMQ:
    url: amqp://localhost
    retry: true

  api:
    baseUrl: http://localhost:3004
    port: 3004

  webapp:
    baseUrl: http://localhost:3000

  smtp:
    host: localhost
    port: 1025
    user:
    password:
    tls: false

  auth:
    code:
      secret: CHANGE ME!!!!!!!!
      expiry: 2h

    token:
      secret: CHANGE ME!!!!!!
      expiry: 4h

    refreshToken:
      secret: CHANGE ME!!!!!!
      expiry: 30d
