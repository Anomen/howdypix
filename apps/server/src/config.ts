// Load env variables
require("dotenv").config();

const config = {
  rabbitMq: {
    url: process.env.RABBITMQ_URL || "amqp://localhost"
  },
  serverApi: {
    baseUrl: "",
    port: parseInt(process.env.API_PORT as string) || 3004
  },
  auth: {
    code: {
      secret: "secret_code_string",
      expiry: "2h"
    },
    token: {
      secret: "secret_token_string",
      expiry: "4h"
    },
    refreshToken: {
      secret: "secret_refresh_string",
      expiry: "30d"
    }
  }
};

config.serverApi.baseUrl =
  process.env.API_BASE_URL || `http://localhost:${config.serverApi.port}`;

export default config;
