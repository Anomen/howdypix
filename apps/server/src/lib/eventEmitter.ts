import StrictEventEmitter from "strict-event-emitter-types";
import { EventEmitter } from "events";
import { ProcessData } from "@howdypix/shared-types";

export interface EventTypes {
  newFile: { root: string; path: string; sourceId: string };
  removeFile: { root: string; path: string; sourceId: string };
  processFile: { root: string; path: string; sourceId: string };
  processedFile: ProcessData;
}

export type Events = StrictEventEmitter<EventEmitter, EventTypes>;
