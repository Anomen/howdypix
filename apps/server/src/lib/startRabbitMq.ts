import { Events } from "./eventEmitter";
import { Channel, connect as connectRabbitMq } from "amqplib";
import {
  HFile,
  MessageProcess,
  ProcessData,
  QueueName
} from "@howdypix/shared-types";
import {
  assertQueue,
  consume,
  sendToQueue,
  appDebug,
  hjoin
} from "@howdypix/utils";
import { UserConfigState } from "../state";

const debug = appDebug("rabbit");

export async function fetchPathsInQueue(channel: Channel): Promise<string[]> {
  const pathsInQueue: string[] = [];

  await consume<MessageProcess>(
    channel,
    QueueName.TO_PROCESS,
    msg => {
      if (msg) {
        pathsInQueue.push(hjoin(msg.data.hfile) as string);
      }
    },
    { consumerTag: "server" }
  );

  await channel.cancel("server");
  await channel.recover();

  debug(`Paths already in queue: ${pathsInQueue.length}.`);
  debug(pathsInQueue);

  return pathsInQueue;
}

export async function bindAppEvents(
  event: Events,
  thumbnailsDir: string,
  channel: Channel
) {
  const pathsInQueue = await fetchPathsInQueue(channel);

  // When there is a new file, send it to the queue
  event.on("processFile", ({ root, hfile }) => {
    const hpath = hjoin(hfile);
    if (pathsInQueue.filter(p => p === hpath).length === 0) {
      appDebug("sendToQueue")(hpath);
      sendToQueue<MessageProcess>(channel, QueueName.TO_PROCESS, {
        thumbnailsDir,
        root,
        hfile
      });
    }
  });
}

export async function bindChannelEvents(event: Events, channel: Channel) {
  await consume<ProcessData>(channel, QueueName.PROCESSED, msg => {
    if (msg) {
      event.emit("processedFile", msg.data);
      channel.ack(msg);
    }
  });
}

export async function startRabbitMq(
  event: Events,
  userConfig: UserConfigState,
  url: string
) {
  try {
    const connection = await connectRabbitMq(url);
    const channel = await connection.createChannel();

    await assertQueue(channel, QueueName.TO_PROCESS);
    await assertQueue(channel, QueueName.PROCESSED);

    await bindAppEvents(event, userConfig.thumbnailsDir, channel);
    await bindChannelEvents(event, channel);

    return channel;
  } catch (e) {
    if (process.env.MOCK) {
      debug(
        "Connecting to RabbitMQ failed, but it's probably normal because you're testing."
      );
    } else {
      throw e;
    }
  }
}
