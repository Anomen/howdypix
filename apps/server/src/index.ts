import "reflect-metadata";
import config from "./config";
import { applyApolloMiddleware } from "./middleware/apollo";
import { loadUserConfig } from "./lib/loadUserConfig";
import { startFileScan } from "./lib/startFileScan";
import EventEmitter from "events";
import { Events } from "./lib/eventEmitter";
import { startRabbitMq } from "./lib/startRabbitMq";
import { startCacheDB } from "./lib/startCacheDB";
import express from "express";
import { staticHandler } from "./middleware/static";
import { emailListHandler, emailViewHandler } from "./middleware/email";
import { applyAuthMiddleware } from "./middleware/auth";

async function main() {
  const event: Events = new EventEmitter();

  const userConfig = loadUserConfig();
  console.log("User Configuration loaded:");
  console.log(userConfig);

  /**
   * Start the mandatory services
   */
  await startCacheDB(event, userConfig);
  await startRabbitMq(event, userConfig, config.rabbitMq.url);
  await startFileScan(event, userConfig);

  /**
   * Start the API
   */
  const app = express();

  app.use(express.json());

  app.get("/static/*", staticHandler);

  // Only when we develop: it's easier to check the email templates
  // and develop them.
  app.get("/email", emailListHandler);
  app.get("/email/*", emailViewHandler);

  // Authentication routes
  applyAuthMiddleware(app);

  // Attach the Apollo middlewares
  applyApolloMiddleware(app, userConfig);

  app.listen({ port: config.serverApi.port }, () => {
    console.log(`API server started on port ${config.serverApi.port}.`);
  });
}

main();
