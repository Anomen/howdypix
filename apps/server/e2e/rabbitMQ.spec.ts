import { HFile, QueueName } from "@howdypix/shared-types";
import { consume, sendToQueue } from "@howdypix/utils";
import { connectToRabbitMq } from "@howdypix/utils/dist/rabbitMq";
import { Channel, Connection } from "amqplib";
import { EventEmitter } from "events";
import { startRabbitMq } from "../src/lib/startRabbitMq";
import { Events, EventTypes } from "../src/lib/eventEmitter";
import { appConfig, UserConfig } from "../src/config";

describe("startRabbitMq", () => {
  let connection: Connection;
  let channel: Channel;
  let events: Events;

  const baseUserConfig: UserConfig = {
    photoDirs: {
      first: "./first",
      second: "./second",
    },
    thumbnailsDir: "/thumbnailsDir",
    users: [],
    emailSender: {
      email: "email@sender.com",
      name: "Sender",
    },
  };

  const baseHfile: HFile = {
    dir: "dir",
    file: "file",
    source: "source",
    name: "name",
  };

  const countMessages = async (queue: QueueName): Promise<number> =>
    (await channel.assertQueue(queue)).messageCount;

  const testSendMessageAgainstSnapshot = async <T extends keyof EventTypes>(
    eventName: T,
    data: EventTypes[T],
    queueName: QueueName,
    numberOfMessagesExpected: number
  ): Promise<void> =>
    new Promise((resolve) => {
      events.once(eventName, async () => {
        setTimeout(async () => {
          expect(await countMessages(queueName)).toEqual(
            numberOfMessagesExpected
          );

          consume(
            channel,
            queueName,
            (msg) => {
              expect(msg?.data).toMatchSnapshot();
            },
            { consumerTag: "consume" }
          )
            .then(() => channel.cancel("consume"))
            .then(() => channel.recover())
            .then(() => resolve());
        }, 100);
      });

      // @ts-ignore
      events.emit(eventName, data);
    });

  beforeAll(async () => {
    connection = await connectToRabbitMq(appConfig.rabbitMQ.url);
    channel = await connection.createChannel();

    // Remove all queues
    try {
      await channel.deleteQueue(QueueName.TO_PROCESS);
      await channel.deleteQueue(QueueName.PROCESSED);
    } catch (e) {
      // do nothing
    }

    // Create new queues
    try {
      await channel.assertQueue(QueueName.TO_PROCESS);
      await channel.assertQueue(QueueName.PROCESSED);
    } catch (e) {
      // do nothing
    }
  });

  beforeEach(async () => {
    events = new EventEmitter();

    try {
      await channel.purgeQueue(QueueName.PROCESSED);
      await channel.purgeQueue(QueueName.TO_PROCESS);
    } catch (e) {
      // do nothing
    }
  });

  afterAll(async () => {
    await channel.close();
    await connection.close();
  });

  it("should send an event when receiving a RabbitMQ message", (done) => {
    // 1. Configure
    events.on("processedFile", (args) => {
      // 3. Test
      expect(args).toMatchSnapshot();
      done();
    });

    // 2. Act
    startRabbitMq(events, baseUserConfig, appConfig.rabbitMQ).then(() => {
      sendToQueue(channel, QueueName.PROCESSED, {
        root: "root",
        hfile: baseHfile,
        exif: {
          aperture: 1,
          createDate: 2,
          ISO: 3,
          make: "make",
          model: "model",
          shutter: 4,
        },
        stat: {
          birthtime: 1,
          ctime: 2,
          inode: 3,
          mtime: 4,
          size: 5,
        },
        thumbnails: [],
      });
    });
  });

  it("should send a rabbitMQ message when receiving `processFile` event", async () => {
    // ==============================
    // 1. Initialize
    // ==============================
    await startRabbitMq(events, baseUserConfig, appConfig.rabbitMQ);

    // ==============================
    // 2. Test
    // ==============================
    expect(await countMessages(QueueName.TO_PROCESS)).toEqual(0);

    await testSendMessageAgainstSnapshot(
      "processFile",
      { root: "/root", hfile: baseHfile },
      QueueName.TO_PROCESS,
      1
    );

    await testSendMessageAgainstSnapshot(
      "processFile",
      { root: "/root", hfile: { ...baseHfile, file: "file2" } },
      QueueName.TO_PROCESS,
      2
    );
  });

  it("should not send the same message twice if already in queue", async () => {
    // ==============================
    // 1. Configure
    // ==============================
    // We send the message to simulate that it's already in the queue
    sendToQueue(channel, QueueName.TO_PROCESS, {
      thumbnailsDir: "thumbnailsDir",
      root: "root",
      hfile: baseHfile,
    });

    await startRabbitMq(events, baseUserConfig, appConfig.rabbitMQ);

    // ==============================
    // 3. Test
    // ==============================
    await testSendMessageAgainstSnapshot(
      "processFile",
      { root: "/root", hfile: baseHfile },
      QueueName.TO_PROCESS,
      1
    );
  });
});
