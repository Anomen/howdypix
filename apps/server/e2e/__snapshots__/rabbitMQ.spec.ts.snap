// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`startRabbitMq should not send the same message twice if already in queue 1`] = `
Object {
  "hfile": Object {
    "dir": "dir",
    "file": "file",
    "name": "name",
    "source": "source",
  },
  "root": "root",
  "thumbnailsDir": "thumbnailsDir",
}
`;

exports[`startRabbitMq should send a rabbitMQ message when receiving \`processFile\` event 1`] = `
Object {
  "hfile": Object {
    "dir": "dir",
    "file": "file",
    "name": "name",
    "source": "source",
  },
  "root": "/root",
  "thumbnailsDir": "/thumbnailsDir",
}
`;

exports[`startRabbitMq should send a rabbitMQ message when receiving \`processFile\` event 2`] = `
Object {
  "hfile": Object {
    "dir": "dir",
    "file": "file",
    "name": "name",
    "source": "source",
  },
  "root": "/root",
  "thumbnailsDir": "/thumbnailsDir",
}
`;

exports[`startRabbitMq should send a rabbitMQ message when receiving \`processFile\` event 3`] = `
Object {
  "hfile": Object {
    "dir": "dir",
    "file": "file2",
    "name": "name",
    "source": "source",
  },
  "root": "/root",
  "thumbnailsDir": "/thumbnailsDir",
}
`;

exports[`startRabbitMq should send an event when receiving a RabbitMQ message 1`] = `
Object {
  "hfile": Object {
    "dir": "dir",
    "file": "file",
    "name": "name",
    "source": "source",
  },
  "root": "root",
  "thumbnailsDir": "thumbnailsDir",
}
`;
